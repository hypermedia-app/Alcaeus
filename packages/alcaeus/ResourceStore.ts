import type { BaseQuad, DatasetCore, NamedNode, Quad } from '@rdfjs/types'
import type { RdfResource, ResourceIdentifier } from '@tpluscode/rdfine'
import type { ResourceFactory } from '@tpluscode/rdfine/lib/ResourceFactory'
import type { RdfResourceCore } from '@tpluscode/rdfine/RdfResource'
import type { AnyContext, AnyPointer, GraphPointer } from 'clownface'
import deleteMatch from 'rdf-dataset-ext/deleteMatch.js'
import addAll from 'rdf-dataset-ext/addAll.js'
import type { CachedResourceFactory } from 'alcaeus-model/ResourceFactory.js'
import CachedResourceFactoryImpl from 'alcaeus-model/ResourceFactory.js'
import type { HydraEnvironment, HydraResponse, ResourceStore, ResourceStoreEntry, ResponseWrapper } from './alcaeus.js'
import ResourceRepresentationImpl from './ResourceRepresentation.js'

export interface RepresentationInference {
  (dataset: DatasetCore, env: HydraEnvironment): Iterable<BaseQuad>
}

interface ResourceStoreInit<D extends DatasetCore> {
  dataset: D
  environment: HydraEnvironment<D>
  inferences: RepresentationInference[]
  factory?: ResourceFactory<D, RdfResource<D>>
}

export default class ResourceStoreImpl<D extends DatasetCore> implements ResourceStore<D> {
  private readonly dataset: D
  private readonly inferences: RepresentationInference[]
  public readonly factory: CachedResourceFactory<D, RdfResource<D>>
  private readonly rootNodes
  private readonly responses
  private readonly environment: HydraEnvironment<D>

  public constructor({ dataset, inferences, factory, environment }: ResourceStoreInit<D>) {
    this.dataset = dataset
    this.inferences = inferences
    if (!factory) {
      factory = environment.rdfine().factory
    }
    this.factory = factory instanceof CachedResourceFactoryImpl ? factory : new CachedResourceFactoryImpl(factory, environment)
    this.environment = environment
    this.rootNodes = environment.termMap<NamedNode, ResourceIdentifier>()
    this.responses = environment.termMap<NamedNode, ResponseWrapper>()
  }

  public clone(): ResourceStore<D> {
    return new ResourceStoreImpl({
      inferences: this.inferences,
      dataset: this.environment.dataset([...this.dataset]),
      factory: this.factory.clone(),
      environment: this.environment,
    })
  }

  public get<T extends RdfResourceCore<D>>(idOrPointer: NamedNode | GraphPointer<ResourceIdentifier, D>): Required<HydraResponse<D, T>> | undefined {
    let graph: NamedNode
    let node: AnyPointer<AnyContext, D>
    if ('termType' in idOrPointer) {
      graph = idOrPointer
      node = this.environment.clownface({ dataset: this.dataset, graph })
    } else {
      graph = idOrPointer._context[0].graph as any
      node = idOrPointer
    }

    const response = this.responses.get(graph)

    if (!response) {
      return undefined
    }

    const rootNode = this.rootNodes.get(graph) || graph
    return {
      response,
      representation: new ResourceRepresentationImpl<D, T>(node, this.environment, rootNode),
    }
  }

  public async set(graph: NamedNode, { response, dataset, rootResource }: ResourceStoreEntry<D>): Promise<void> {
    this.factory.removeCache(graph)

    const inferredQuads = this.environment.dataset()

    this.inferences.forEach(inferenceOver => addAll(inferredQuads, [...inferenceOver(dataset, this.environment)]))

    addAll(
      deleteMatch(this.dataset, undefined, undefined, undefined, graph),
      [...dataset].map(this.tripleToQuad(graph)),
    )
    addAll(this.dataset, [...inferredQuads].map(this.tripleToQuad(graph)))

    this.responses.set(graph, response)

    if (rootResource) {
      this.rootNodes.set(graph, rootResource)
    }
  }

  private tripleToQuad(graph: NamedNode) {
    return (triple: Quad): Quad => {
      return this.environment.quad(triple.subject, triple.predicate, triple.object, graph)
    }
  }
}
