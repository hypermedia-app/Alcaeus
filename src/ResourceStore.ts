import { RdfResource, ResourceFactory } from '@tpluscode/rdfine'
import cf from 'clownface'
import TripleToQuadTransform from 'rdf-transform-triple-to-quad'
import { DatasetIndexed } from 'rdf-dataset-indexed/dataset'
import { DatasetCore, NamedNode, BaseQuad } from 'rdf-js'
import $rdf from '@rdfjs/data-model'
import ResourceRepresentationImpl, { ResourceRepresentation } from './ResourceRepresentation'
import { HydraResource } from './Resources'

export interface ResourceStore<D extends DatasetIndexed> {
    factory: ResourceFactory<D, HydraResource<D>>
    get<T extends RdfResource = RdfResource>(uri: string | NamedNode): ResourceRepresentation<T> | undefined
    set(uri: string | NamedNode, dataset: D, rootResource?: NamedNode): Promise<void>
    clone(): ResourceStore<D>
}

export interface RepresentationInference {
    (dataset: DatasetCore): Iterable<BaseQuad>
}

interface ResourceStoreInit<D extends DatasetIndexed = DatasetIndexed> {
    dataset: D
    datasetFactory: () => D
    inferences: RepresentationInference[]
    factory: ResourceFactory<D, HydraResource<D>>
}

export default class ResourceStoreImpl<D extends DatasetIndexed = DatasetIndexed> implements ResourceStore<D> {
    private readonly dataset: D;
    private readonly inferences: RepresentationInference[];
    public readonly factory: ResourceFactory<D, HydraResource<D>>
    private readonly rootNodes: Map<string, NamedNode> = new Map<string, NamedNode>()
    private readonly datasetFactory: () => D;

    public constructor({ dataset, inferences, factory, datasetFactory }: ResourceStoreInit<D>) {
        this.dataset = dataset
        this.inferences = inferences
        this.factory = factory
        this.datasetFactory = datasetFactory
    }

    public clone(): ResourceStore<D> {
        return new ResourceStoreImpl({
            inferences: this.inferences,
            dataset: this.dataset.clone(),
            factory: this.factory,
            datasetFactory: this.datasetFactory,
        })
    }

    public get<T extends RdfResource>(uri: string | NamedNode): ResourceRepresentation<T> | undefined {
        const graph = typeof uri === 'string' ? $rdf.namedNode(uri) : uri
        const node = cf({ dataset: this.dataset, graph })

        if (!node.out().values.length) {
            return undefined
        }

        return new ResourceRepresentationImpl<T>(node, this.factory, this.rootNodes.get(graph.value) || graph)
    }

    public async set(uri: string | NamedNode, dataset: D, rootResource?: NamedNode): Promise<void> {
        const graph = typeof uri === 'string' ? $rdf.namedNode(uri) : uri
        const inferredQuads = this.datasetFactory()

        this.inferences.forEach(inferenceOver => inferredQuads.addAll([...inferenceOver(dataset)]))

        await this.dataset
            .removeMatches(undefined, undefined, undefined, graph)
            .import(dataset.toStream().pipe(new TripleToQuadTransform(graph)))
        await this.dataset.import(inferredQuads.toStream().pipe(new TripleToQuadTransform(graph)))

        if (rootResource) {
            this.rootNodes.set(graph.value, rootResource)
        }
    }
}
