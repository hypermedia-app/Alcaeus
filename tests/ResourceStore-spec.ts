import { rdf } from '@tpluscode/rdf-ns-builders'
import RdfResource from '@tpluscode/rdfine'
import DatasetExt from 'rdf-ext/lib/Dataset'
import $rdf from 'rdf-ext'
import * as RDF from '@rdf-esm/data-model'
import namespace from '@rdfjs/namespace'
import ResourceStore, { RepresentationInference } from '../src/ResourceStore'

const ex = namespace('http://example.com/')

describe('ResourceStore', () => {
    const factory = RdfResource.factory
    const datasetFactory = $rdf.dataset
    let dataset: DatasetExt
    let inferences: RepresentationInference[]

    beforeEach(() => {
        dataset = $rdf.dataset()
        inferences = []
    })

    describe('get', () => {
        it('returns undefined when resource is not stored', () => {
            // given
            const store = new ResourceStore({
                dataset,
                inferences,
                factory,
                datasetFactory,
            })

            // when
            const resource = store.get(ex.foo)

            // then
            expect(resource).toBeUndefined()
        })
    })

    describe('set', () => {
        it('add inferred triples to same graph', async () => {
            // given
            const addTypeInference = () => {
                return [
                    RDF.quad(ex.foo, rdf.type, ex.baz, ex.graphShouldBeDiscareded),
                ]
            }
            const store = new ResourceStore({
                dataset,
                inferences: [addTypeInference],
                factory,
                datasetFactory,
            })
            const resourceDataset = $rdf.dataset([
                RDF.quad(ex.foo, rdf.type, ex.bar, ex.foo),
            ])

            // when
            await store.set(ex.foo, resourceDataset)

            // then
            expect(dataset.toCanonical()).toMatchSnapshot()
        })

        it('add resource triples to graph as subject', async () => {
            // given
            const store = new ResourceStore({
                dataset,
                inferences,
                factory,
                datasetFactory,
            })
            const resourceDataset = $rdf.dataset([
                RDF.quad(ex.foo, rdf.type, ex.bar, ex.whatever),
            ])

            // when
            await store.set(ex.foo, resourceDataset)

            // then
            expect(dataset.toCanonical()).toMatchSnapshot()
        })
    })
})
