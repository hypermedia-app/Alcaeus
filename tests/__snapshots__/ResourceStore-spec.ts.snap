// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceStore set add inferred triples to same graph 1`] = `
"<http://example.com/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/bar> <http://example.com/foo> .
<http://example.com/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/baz> <http://example.com/foo> .
"
`;

exports[`ResourceStore set add resource triples to graph as subject 1`] = `
"<http://example.com/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://example.com/bar> <http://example.com/foo> .
"
`;
